cypress/integration/spec.js

```diff
@@ -76,31 +88,90 @@ describe('Chart Explorer', function() {
     ).click()
     cy.get('[data-test=rm-encoding]').click()
 
-    // check for editor existence
-    cy.get('#configure-tabs-tab-editor').click()
-    cy.get('.monaco-editor')
-    cy.get('#configure-tabs-tab-builder').click()
-
     // check that titles work
     cy.get('[data-test=chart-title]').type('test title')
     cy.get('svg .role-title').contains('test title')
 
     // check that chart exists
     cy.get('[data-test=vega-embed]')
-
     cy.screenshot('page with viz')
 
-    // check download options
+    cy.get('[data-test=license-open]').click()
+    cy.get('[data-test=license-text]').contains('cypress license text')
+  })
+
+  it('Should be able to edit chart using editor', function() {
+    cy.get('#configure-tabs-tab-editor').click()
+
+    cy.get('.inputarea')
+      .type('{ctrl}f')
+      .focused()
+      .type('"title": "test 123",')
+
+    cy.get('#configure-tabs-tab-builder').click()
+
+    cy.get('svg .role-title').contains('test 123')
+  })
+
+  it('Should be able to download a chart', function() {
     cy.get('#dropdown-download').click()
 
+    // download as a vega-lite file format
     cy.get('.open > .dropdown-menu > :nth-child(3) > a').trigger('mousedown')
     cy.get('.open > .dropdown-menu > :nth-child(3) > a').should($m => {
       expect($m).to.have.length(1)
       expect($m.attr('href')).to.match(/^blob/)
     })
-    cy.get('#dropdown-download').click()
 
-    // check download as insight
+    // download as a vega file format
+    cy.get('.open > .dropdown-menu > :nth-child(2) > a').trigger('mousedown')
+    cy.get('.open > .dropdown-menu > :nth-child(2) > a').should($m => {
+      expect($m).to.have.length(1)
+      expect($m.attr('href')).to.match(/^blob/)
+    })
+
+    // download as a png file format
+    cy.get('.open > .dropdown-menu > :nth-child(5) > a').trigger('mousedown')
+    cy.get('.open > .dropdown-menu > :nth-child(5) > a').should($m => {
+      expect($m).to.have.length(1)
+      expect($m.attr('href')).to.match(/^data:image\/png/)
+    })
+
+    // download as a svg file format
+    cy.get('.open > .dropdown-menu > :nth-child(6) > a').trigger('mousedown')
+    cy.get('.open > .dropdown-menu > :nth-child(6) > a').should($m => {
+      expect($m).to.have.length(1)
+      expect($m.attr('href')).to.match(/^blob/)
+    })
+
+    // download as a HTML file format
+    cy.get('.open > .dropdown-menu > :nth-child(8) > a').trigger('mousedown')
+    cy.get('.open > .dropdown-menu > :nth-child(8) > a').should($m => {
```

Comment:
I really don't think that `:nth-child(...)` is a good pattern since:
- DOM elements might be reordered in future changes, and should have it's own separate test (maybe a snapshot test is sufficient?)
- it's difficult to read without having the app running and checking which element you're checking at every index and DOM traversal should be abstracted

Is it possible to use `data-test` tags instead?

Reply:
I agree that using `nth-child` doesn't feel like the best approach here. It feels to me like `nth-child` is only really a good idea to use when you're testing something specific to an element in the nth position, regardless of the content. but if you're trying to target an element of specific content, it feels best to use the content as the means to fetch it (like with a `data-dw` tag, which specifically denotes certain content. does this make sense?